### ISSUE TITLE
The specified CGI application encountered an error and the server terminated the process.

### ISSUE THREAD

**Issue by sherif-elmetainy** [dotnet/aspnetcore#18735]
I have an ASP.NET core 2.2 application hosted on Azure, and I encountered today an error:
The specified CGI application encountered an error and the server terminated the process.

Looking at the logs, I saw the exception details. It was a service that calls an async method in the Dispose() method, and the async method threw System.Data.SqlClient.SqlException: Timeout expired. 

Since this error occurred "outside of a context", it was not caught by any middleware or parts of the framework and the application was terminated. This is a bug in my code that is easy to fix. 

All the above is understandable and fine.

The exception was in the logs one time and the AspNetCore Module or HttpPlatformHandler didn't attempt to restart the process again. As a result, the application was down for about an hour until I manually restarted it

The rapidFailuresPerMinute setting is not set, so it should have a default value of 10. However, it seems the actual behaviour is one failure and application is down until restarted.

I do understand that it is a bug on my side that ultimately caused the crash, but such bugs can happen. It would be nice if the Framework logs the error then recovers.

**Comment by anurse** [comment]
> It would be nice if the Framework logs the error then recovers.

We don't capture crashing exceptions in out-of-proc mode because we don't have any way to intercept the crash from a separate process. However, the runtime should be writing crashing exceptions to the Event Log. Are you seeing any exception information in the Windows Event Log?

**Comment by sherif-elmetainy** [comment]
@anurse Thanks for your answer. Yes, I saw the exception with full stacktrace in the eventlog.xml on Azure and this is how I easily found and fixed the bug in my code. So the "logging" part is working fine.  

May be I didn't explain well. I didn't mean intercept (or catch) the exception. I meant monitor the process and if it crashes and dies, start a new process (so the application functionality is restored).

**Comment by anurse** [comment]
> May be I didn't explain well. I didn't mean intercept (or catch) the exception. I meant monitor the process and if it crashes and dies, start a new process (so the application functionality is restored).

This should be happening. If the process crashes (and you don't hit the rapid failure threshold, which you indicated you aren't) it should be restarting. Can you share a runnable sample project that reproduces the issue?

> I have an ASP.NET core 2.2 application hosted on Azure

FYI, ASP.NET Core 2.2 is [end-of-life and no longer supported](https://devblogs.microsoft.com/dotnet/net-core-2-2-will-reach-end-of-life-on-december-23-2019/) you should migrate to either 3.1 (the latest LTS) or 2.1 (the previous LTS) to ensure you maintain support.

**Comment by sherif-elmetainy** [comment]
@anurse Unfortunately I don't have time create a sample project at the moment. I tried a hello world thing with very little code, and that didn't seem to reproduce the problem, so there is something causing it that I don't know of and it needs more time to find. May be I can have time on the next weekend. 
I looked back at the logs and found the following error right after the logged exception:

Failed to gracefully shutdown process with Id '24328'. 

May be this piece of information can help you figure what's going on.

As for .NET Core 3.1, I plan to upgrade next month.
